codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/customizations/UserProvidedValidationExceptionDecorator.kt

```diff
@@ -0,0 +1,620 @@
+import software.amazon.smithy.rust.codegen.core.util.toSnakeCase
+import software.amazon.smithy.rust.codegen.server.smithy.ServerCodegenContext
+import software.amazon.smithy.rust.codegen.server.smithy.customize.ServerCodegenDecorator
+import software.amazon.smithy.rust.codegen.server.smithy.generators.BlobLength
+import software.amazon.smithy.rust.codegen.server.smithy.generators.CollectionTraitInfo
+import software.amazon.smithy.rust.codegen.server.smithy.generators.ConstraintViolation
+import software.amazon.smithy.rust.codegen.server.smithy.generators.Range
+import software.amazon.smithy.rust.codegen.server.smithy.generators.StringTraitInfo
+import software.amazon.smithy.rust.codegen.server.smithy.generators.UnionConstraintTraitInfo
+import software.amazon.smithy.rust.codegen.server.smithy.generators.ValidationExceptionConversionGenerator
+import software.amazon.smithy.rust.codegen.server.smithy.generators.isKeyConstrained
+import software.amazon.smithy.rust.codegen.server.smithy.generators.isValueConstrained
+import software.amazon.smithy.rust.codegen.server.smithy.generators.protocol.ServerProtocol
+import software.amazon.smithy.rust.codegen.server.smithy.util.isValidationMessage
+import software.amazon.smithy.rust.codegen.server.smithy.validationErrorMessage
+import software.amazon.smithy.rust.codegen.traits.ValidationExceptionTrait
+import software.amazon.smithy.rust.codegen.traits.ValidationFieldListTrait
+import software.amazon.smithy.rust.codegen.traits.ValidationFieldMessageTrait
+import software.amazon.smithy.rust.codegen.traits.ValidationFieldNameTrait
+
+/**
+ * Decorator for custom validation exception codegen
+ *
+ * The order of this is less than that of [SmithyValidationExceptionDecorator] so it takes precedence regardless of the
+ * order decorators are passed into the plugin
+ */
+class UserProvidedValidationExceptionDecorator : ServerCodegenDecorator {
+    override val name: String
+        get() = "CustomValidationExceptionDecorator"
+    override val order: Byte
+        get() = 68
+
+    internal fun customValidationException(codegenContext: ServerCodegenContext): StructureShape? =
+        codegenContext.model
+            .shapes(StructureShape::class.java)
+            .filter { it.hasTrait(ValidationExceptionTrait.ID) }
+            .findFirst()
+            .orElse(null)
+
+    override fun validationExceptionConversion(
+        codegenContext: ServerCodegenContext,
+    ): ValidationExceptionConversionGenerator? {
+        val validationException = customValidationException(codegenContext) ?: return null
+
+        return UserProvidedValidationExceptionConversionGenerator(codegenContext, validationException)
+    }
+}
+
+class UserProvidedValidationExceptionConversionGenerator(
+    private val codegenContext: ServerCodegenContext,
+    private val validationException: StructureShape,
+) : ValidationExceptionConversionGenerator {
+    private val maybeValidationField = customValidationField()
```

Comment:
Can we make this a lazy property? Otherwise it executes immediately when the object is created:

```kotlin
private val maybeValidationField by lazy {
    customValidationField()
}
```